Cleanup

Crunchwrap78 opened the pull request:
Things to note.
- eslint needs to be updated for es6/es7 syntax in classes
- vendor file before compression is still big in size. vendor compressed resizes well, only thing is the size is still pushing the limit recommended
- dependencies were updated in the process.
- I'll be leaving comments with explanations on changes for clarity

Crunchwrap78 committed:
updated dependencies and refactored configuration

Crunchwrap78 committed:
part 1 of refactor

Crunchwrap78 committed:
webpack configure & fetch and main page refactored

Crunchwrap78 committed:
refactored members and teams. Need to fix member profile bug

Crunchwrap78 committed:
refactor complete

Crunchwrap78 committed:
fully updated webpack 4

Crunchwrap78 commented on .babelrc:
```diff
@@ -1,6 +1,19 @@
+      }
+    ],
     "react"
+  ],
```
these plugins allow for us to use 
1) arrow functions for functions, so no more setting the context of this. 
2) spread operator, so you use the spread syntax in our classes
3) dynamic imports, so you can now dynamically import dependencies instead of how we module import them in each file. to note this is used for code splitting
4) transform runtime is optional, but references all the helpers babel uses reducing duplicate helpers.

Crunchwrap78 committed:
forgot to add props to title

Crunchwrap78 committed:
quick fix

Crunchwrap78 committed:
changes?

Crunchwrap78 committed:
cleanup

keithb418 commented on front-end/src/actions/fetch.js:
```diff
@@ -0,0 +1,15 @@
+
+export const fetchDataError = () => ({
+    type: types.FETCH_DATA_ERROR,
+    error: "An error has occur"
```
Typo: An error has occurred

keithb418 commented on front-end/src/actions/shared.js:
```diff
@@ -0,0 +1,34 @@
+
+export const fetchInitialData = () => dispatch => {
+    dispatch(fetchData());
+    Promise.all([API.getRoles(), API.getTeams(), API.getTeamMemebers()])
```
typo: getTeamMembers

keithb418 commented on front-end/src/actions/shared.js:
```diff
@@ -0,0 +1,34 @@
+
+export const fetchProfileData = id => dispatch => {
+   dispatch(fetchData());
+   Promise.all([API.getTeamMemeber(id), API.getTeams()])
```
Typo: getTeamMember

keithb418 commented on front-end/src/actions/teamMembers.js:
```diff
@@ -1,126 +1,68 @@
-  _createTeamMember(teamMember)(dispatch)
+export const handleFetchTeamMember = id => dispatch => {
+	dispatch(fetchData())
+	API.getTeamMemeber(id)
```
typo: getTeamMember

keithb418 commented on front-end/src/actions/teamMembers.js:
```diff
@@ -1,126 +1,68 @@
+})
+
+
+export const createTeamMember = (teamMember) => (dispatch) => {
```
Rename to handleCreateTeamMember for consistency with handleFetchTeamMember.  It'll show the difference between the redux action to be dispatched and the function to be called which actually makes the request to the API.

keithb418 commented on front-end/src/actions/teamMembers.js:
```diff
@@ -1,126 +1,68 @@
+	)
+}
+
+export const hanldeUpdateTeamMember = (teamMember) => (dispatch) => {
```
typo: handleUpdateTeamMember

keithb418 commented on front-end/src/actions/teamMembers.js:
```diff
@@ -1,126 +1,68 @@
+
+export const deleteTeamMember = (id) => (dispatch) => {
+	dispatch({ type: types.DELETE_TEAM_MEMBER, id })
+	API.deleteTeamMemebers(id)
```
typo: deleteTeamMembers

keithb418 commented on front-end/src/actions/teamMembers.js:
```diff
@@ -1,126 +1,68 @@
+	);	 
+}
+
+export const deleteTeamMember = (id) => (dispatch) => {
```
Update to handleDeleteTeamMember for consistency.

keithb418 commented on front-end/src/actions/teamMembers.js:
```diff
@@ -1,126 +1,68 @@
+	);
+}
+
+export const changeTeam = (_id, team) => (dispatch) => {
```
update to handleChangeTeam for consistency

keithb418 commented on front-end/src/actions/teams.js:
```diff
@@ -1,96 +1,47 @@
\ No newline at end of file
+}
+
+export const deleteTeam = id => dispatch => {
```
rename to handleDeleteTeam for consistency

keithb418 commented on front-end/src/actions/teams.js:
```diff
@@ -1,96 +1,47 @@
+    .catch(err => console.log('Could not delete team:', err.message));
+}
+    
+export const updateTeam = (team) => dispatch => {
```
rename to handleUpdateTeam for consistency

keithb418 commented on front-end/src/actions/teams.js:
```diff
@@ -1,96 +1,47 @@
+    });
+}
+
+export const createTeam = team => dispatch =>{
```
change to handleCreateTeam for consistency

keithb418 commented on front-end/src/components/App.js:
```diff
@@ -1,39 +1,77 @@
+import Loading from './Loading';
+import DynamicImport from './DynamicImport';
+
+const TeamsColumns = (props) => (
```
Each of these components should be separated into its own file.

keithb418 commented on front-end/src/components/Loading.js:
```diff
@@ -2,14 +2,18 @@ import React from "react";
 
-const Loading = ({ isLoading }) =>
-  isLoading &&
+const Loading = (props) => {
```
We can remove props if it's not needed.

keithb418 commented on front-end/src/utils/api.js:
```diff
@@ -0,0 +1,54 @@
+        this._api = axios.create({ baseURL: "/api" });
+    }
+
+    getTeamMemeber (id) {
```
typo: getTeamMember

keithb418 commented on front-end/src/utils/api.js:
```diff
@@ -0,0 +1,54 @@
+        return this._api.get("roles");
+    }
+
+    getTeamMemebers () {
```
typo: getTeamMembers

keithb418 commented on front-end/src/utils/api.js:
```diff
@@ -0,0 +1,54 @@
+        return this._api.put("teams", team);
+    }
+
+    deleteTeamMemebers (id) {
```
typo: deleteTeamMembers

keithb418 reviewed (changes requested):
In general, it looks good.  Overall some feedback:

1. Watch the typos, I found several
2. We should keep components in separate files, I noticed we had a whole bunch of dynamic loader components in App.js, if possible we should have them in separate files, it would be easier to maintain
3. There should be some sort of differentiation between the actions that you just dispatch to update the store, and the actions that actually interact with the API.  I put down some possibilities, but feel free to come up with your own solution.

And a couple clarifying questions (just out of curiosity):

1. Why was React Router removed?
2. What is the purpose of the StateLoader / why are we saving things to localStorage?

Crunchwrap78 commented:
@keithb418 
1) will fix
2) I addressed this in the comments
3) will fix

other questions
1) I thought I posted that comment but I guess not. react-router-redux  is unnecessary in react-router-dom since it internally handles your history. if you were using the older version before version 4 then yes you would need it.
2) with redux your store is never persisted so you always have to reload your data. There are other remedies to solve this in react like graphql + apollo or relay to bind the data to components. This will speed up performance in loading our app from different views since our props are always persisted. I will note I'm not a complete expert at this seeing as my state structure is pretty basic, but I understand the logistics of it. In general we should only use the data needed in our components, but I added everything in state since in the future the structure will change. I haven't seen any articles that have said you shouldn't do this and actually some encouraged it. You can actually view twitter's state in the local storage

Crunchwrap78 committed:
fixed typos and made actions more consistant

Crunchwrap78 committed:
fixed typo

Crunchwrap78 committed:
resolved typos

asc2683 commented:
This is more of a general comment to do with project branch naming conventions. Since this is more of a major refactoring task, perhaps consider naming the branch as refactor/release-1.1

See contributing doc for branch naming convention: https://github.com/keithb418/scrum-team-app/blob/development/CONTRIBUTING.md

Also, for future refactoring, breaking the refactoring to small tasks into specific modules or functionality (i.e., webpact, react-router, or redux state manegement) would be helpful.

Crunchwrap78 commented:
@asc2683 The react-router and webpack refactor went hand in hand with the dynamic imports. The only thing that should have been separated was the redux state management. After this merges I'll get started on unit tests

Crunchwrap78 committed:
Update README.md

Crunchwrap78 committed:
Update README.md

Crunchwrap78 committed:
Update README.md

Crunchwrap78 committed:
Update README.md

Crunchwrap78 committed:
Update README.md

Crunchwrap78 committed:
Update README.md
